### PR TITLE
Iaco 678 sdk to return user data to merchant

### PR DESCRIPTION
**Pull request by yuval-netanel**

### CODE REVIEW

File: KlarnaOnDemand/KODRegistrationViewController.m [hunk 1]
```diff
@@ -52,14 +52,26 @@ - (void)dismissButtonPressed {
 }
 
 - (void)handleUserReadyEventWithPayload:(NSDictionary *)payload {
-  NSString *token = payload[@"userToken"];
+  NSMutableDictionary *mutablePayload = [payload mutableCopy];
+  NSString *token = [self popFromDictionary:mutablePayload withKey:@"userToken"];
   NSAssert(token, @"KODToken failed to create.");
```
Comment (tomerlevi444) [inline KlarnaOnDemand/KODRegistrationViewController.m]: `KODToken failed to create.` -> `Failed to create KODToken.`

File: KlarnaOnDemand/KODRegistrationResult.h [hunk 2]
```diff
@@ -10,6 +10,16 @@
...
 @property (strong, nonatomic) NSString *token;
 
-- (id)initWithToken:(NSString *)token;
+/**
+ *  The user's validated phone number.
+ */
+@property (strong, nonatomic) NSString *phoneNumber;
+
+/**
+ *  The user's registration details.
```
Comment (grozen) [inline KlarnaOnDemand/KODRegistrationResult.h]: This is really too vague. At the very least we need some link to some docs that don't exist. Talk to the Goor to iron this out, won't you?

File: KlarnaOnDemand/KODRegistrationViewController.m (hunk 1 above)
Comment (grozen) [inline KlarnaOnDemand/KODRegistrationViewController.m]: :+1:

File: KlarnaOnDemandTests/KODRegistrationResultSpec.m [hunk 3]
```diff
@@ -4,15 +4,27 @@
...
    
    it(@"should return false when tokens are not equal", ^{
-     KODRegistrationResult *resultA =[[KODRegistrationResult alloc] initWithToken:@"same_token"];
-     KODRegistrationResult *resultB =[[KODRegistrationResult alloc] initWithToken:@"not_the_same_token"];
+     KODRegistrationResult *resultA =[[KODRegistrationResult alloc] initWithToken:@"same_token" andPhoneNumber:@"same_phoneNumber" andUserDetails:@{@"a":@1}];
+     KODRegistrationResult *resultB =[[KODRegistrationResult alloc] initWithToken:@"not_the_same_token" andPhoneNumber:@"same_phoneNumber" andUserDetails:@{@"a":@1}];
+     [[theValue([resultA isEqual:resultB]) should] beFalse];
+   });
+   
+   it(@"should return false when phone number are not equal", ^{
```
Comment (grozen) [inline KlarnaOnDemandTests/KODRegistrationResultSpec.m]: `phone number` -> `phone numbers`

File: KlarnaOnDemand/KODRegistrationViewController.m [hunk 4]
```diff
@@ -52,14 +52,26 @@ - (void)dismissButtonPressed {
...
   if ([self.delegate respondsToSelector:@selector(klarnaRegistrationController:finishedWithResult:)]) {
-    KODRegistrationResult *kodRegistrationResult = [[KODRegistrationResult alloc] initWithToken:token];
+    NSString *phoneNumber = [self popFromDictionary:mutablePayload withKey:@"phoneNumber"];
+    KODRegistrationResult *kodRegistrationResult = [[KODRegistrationResult alloc]
+                                                    initWithToken:token
+                                                    andPhoneNumber:phoneNumber
+                                                    andUserDetails:mutablePayload];
```
Comment (tomerlevi444) [inline KlarnaOnDemand/KODRegistrationViewController.m]: Reminder: modify the code to fit the new structure of the payload

File: KlarnaOnDemandTests/KODRegistrationViewControllerSpec.m [hunk 5]
```diff
@@ -15,9 +15,13 @@
   });
   
   it(@"should call delegate's .klarnaRegistrationController:finishedResilt on .handleUserReadyEvent when a token was received", ^{
```
Comment (tomerlevi444) [inline KlarnaOnDemandTests/KODRegistrationViewControllerSpec.m]: `klarnaRegistrationController:finishedResilt` -> `klarnaRegistrationController:finishedWithResult`

File: KlarnaOnDemandTests/KODRegistrationResultSpec.m (hunk 3 above)
Comment (grozen) [inline KlarnaOnDemandTests/KODRegistrationResultSpec.m]: Even better `the phone numbers`

File: KlarnaOnDemandTests/KODRegistrationResultSpec.m [hunk 6]
```diff
@@ -4,15 +4,27 @@
 
 describe(@"KODRegistrationResultSpec", ^{
  describe(@"isEqual", ^{
-   it(@"should return true when tokens are equal", ^{
-     KODRegistrationResult *resultA =[[KODRegistrationResult alloc] initWithToken:@"same_token"];
-     KODRegistrationResult *resultB =[[KODRegistrationResult alloc] initWithToken:@"same_token"];
+   it(@"should return true when tokens, phone numbers and user details are equal", ^{
```
Comment (grozen) [inline KlarnaOnDemandTests/KODRegistrationResultSpec.m]: `when tokens` -> `when the tokens`

File: KlarnaOnDemandTests/KODRegistrationResultSpec.m [hunk 7]
```diff
@@ -4,15 +4,27 @@
...
+     KODRegistrationResult *resultA =[[KODRegistrationResult alloc] initWithToken:@"same_token" andPhoneNumber:@"same_phoneNumber" andUserDetails:@{@"a":@1}];
+     KODRegistrationResult *resultB =[[KODRegistrationResult alloc] initWithToken:@"same_token" andPhoneNumber:@"same_phoneNumber" andUserDetails:@{@"a":@1}];
      [[theValue([resultA isEqual:resultB]) should] beTrue];
    });
    
    it(@"should return false when tokens are not equal", ^{
```
Comment (grozen) [inline KlarnaOnDemandTests/KODRegistrationResultSpec.m]: `when tokens` -> `when the tokens`

File: KlarnaOnDemandTests/KODRegistrationResultSpec.m [hunk 8]
```diff
@@ -4,15 +4,27 @@
...
+     [[theValue([resultA isEqual:resultB]) should] beFalse];
+   });
+   
+   it(@"should return false when phone number are not equal", ^{
+     KODRegistrationResult *resultA =[[KODRegistrationResult alloc] initWithToken:@"same_token" andPhoneNumber:@"same_phoneNumber" andUserDetails:@{@"a":@1}];
+     KODRegistrationResult *resultB =[[KODRegistrationResult alloc] initWithToken:@"same_token" andPhoneNumber:@"not_the_same_phoneNumber" andUserDetails:@{@"a":@1}];
+     [[theValue([resultA isEqual:resultB]) should] beFalse];
+   });
+   
+   it(@"should return false when user details are not equal", ^{
```
Comment (grozen) [inline KlarnaOnDemandTests/KODRegistrationResultSpec.m]: `when user` -> `when the user`

**Comment by grozen** [comment]
Approved!

File: KlarnaOnDemand/KODRegistrationViewController.m (hunk 1 above)
Comment (yuval-netanel) [inline KlarnaOnDemand/KODRegistrationViewController.m]: I removed it.

File: KlarnaOnDemandTests/KODRegistrationResultSpec.m (hunk 3 above)
Comment (yuval-netanel) [inline KlarnaOnDemandTests/KODRegistrationResultSpec.m]: :+1:

File: KlarnaOnDemandTests/KODRegistrationViewControllerSpec.m (hunk 5 above)
Comment (yuval-netanel) [inline KlarnaOnDemandTests/KODRegistrationViewControllerSpec.m]: :+1:

File: KlarnaOnDemandTests/KODRegistrationResultSpec.m (hunk 7 above)
Comment (yuval-netanel) [inline KlarnaOnDemandTests/KODRegistrationResultSpec.m]: :+1:

File: KlarnaOnDemand/KODRegistrationResult.h (hunk 2 above)
Comment (yuval-netanel) [inline KlarnaOnDemand/KODRegistrationResult.h]: We decided that we will do it in the documentation ticket (IACO-684)

File: KlarnaOnDemand/KODRegistrationViewController.m (hunk 4 above)
Comment (yuval-netanel) [inline KlarnaOnDemand/KODRegistrationViewController.m]: :+1:

File: KlarnaOnDemandTests/KODRegistrationResultSpec.m (hunk 6 above)
Comment (yuval-netanel) [inline KlarnaOnDemandTests/KODRegistrationResultSpec.m]: :+1:

File: KlarnaOnDemandTests/KODRegistrationResultSpec.m (hunk 8 above)
Comment (yuval-netanel) [inline KlarnaOnDemandTests/KODRegistrationResultSpec.m]: :+1:

**Comment by tomerlevi444** [comment]
Approved

**Pull request merged**